util.c

```diff
@@ -5227,6 +5227,12 @@ S_mem_log_common(enum mem_log_type mlt, const UV n,
             PERL_UNUSED_RESULT(PerlLIO_write(fd, buf, len));
 #ifdef USE_C_BACKTRACE
             if(strchr(pmlenv,'c') && (mlt == MLT_NEW_SV)) {
+                /* TODO: get caller package in here when we work out how */
```

Comment:
[`CopSTASHPV`](https://perldoc.perl.org/perlapi#CopSTASHPV)?

Reply:
I tried `CopSTASHPV` and it just immediately SEGV'ed. I see other calls to it that apparently work so I'm not sure what I'm doing differently.

Reply:
The memory logging functions can be called really early, for threaded builds before PL_stashpad is initialized.  CopSTASH() is `PL_stashpad[(c)->cop_stashoff]` for threaded builds, so it segfaults.

We could make CopSTASH() check that PL_stashpad is non-NULL, but then every other call which doesn't need that check will also be paying the cost.

Using CopFILE() as a proxy for checking for the validity of CopSTASHPV() seemed to work:
```
diff --git a/util.c b/util.c
index fa2d129186..6069c80cfe 100644
--- a/util.c
+++ b/util.c
@@ -5229,8 +5229,9 @@ S_mem_log_common(enum mem_log_type mlt, const UV n,
             if(strchr(pmlenv,'c') && (mlt == MLT_NEW_SV)) {
                 /* TODO: get caller package in here when we work out how */
                 len = my_snprintf(buf, sizeof(buf),
-                        "  caller at %s line %d\n",
-                        CopFILE(PL_curcop), CopLINE(PL_curcop));
+                        "  caller at %s line %d stash %s\n",
+                        CopFILE(PL_curcop), CopLINE(PL_curcop),
+                        CopFILE(PL_curcop) ? CopSTASHPV(PL_curcop) : "<unknown>");
                 PERL_UNUSED_RESULT(PerlLIO_write(fd, buf, len));
 
                 Perl_c_backtrace *bt = Perl_get_c_backtrace(aTHX_ 3, 3);
```

I didn't see a crash with non-threaded builds.

Reply:
Ahyes, that seems to do it.